test/REXLaunchpad.test.ts

```diff
@@ -0,0 +1,440 @@
+        console.log("====== Transferred to alice =======");
+        await ricochetETHx
+            .transfer({
+                receiver: bobSigner.address,
+                amount: ethers.utils.parseUnits("0.5", 18).toString(),
+            }).exec(ethxWhaleSigner);
+            console.log("ETH")
+        await ricochetRIC
+            .transfer({
+                receiver: bobSigner.address,
+                amount: '1000000000000000000000',
+            }).exec(adminSigner);
+            console.log("RIC")
+        await ricochetMATICx
+            .transfer({
+                receiver: bobSigner.address,
+                amount: '1754897259852523432',
+            }).exec(maticxWhaleSigner);
+            console.log("MATIC")
+        console.log("====== Transferred to bob =======");
+        await ricochetUSDCx
+            .transfer({
+                receiver: karenSigner.address,
+                amount: initialAmount,
+            }).exec(usdcxWhaleSigner);
+        console.log("====== Transferred to karen =======");
+
+        // Take a snapshot to avoid redoing the setup
+        snapshot = await provider.send('evm_snapshot', []);
+
+    });
+
+    context("#1 - rex launchpad check referral", async () => {
+
+        beforeEach(async () => {
+            // Revert to the point REXMarket was just deployed
+            const success = await provider.send('evm_revert', [
+                snapshot
+            ]);
+            // Take another snapshot to be able to revert again next time
+            snapshot = await provider.send('evm_snapshot', []);
+            expect(success).to.equal(true);
+        });
+
+        afterEach(async () => {
+            // Check the app isn't jailed
+            expect(await launchpad.isAppJailed()).to.equal(false);
+            await resetMeasurements();
+        });
+
+
+        it("#1.2 before/afterAgreementCreated callbacks", async () => {
+
+            // Alice opens a USDC stream to launchoad
+            await sf.cfaV1.createFlow({
+                sender: aliceSigner.address,
+                receiver: launchpad.address,
+                superToken: ricochetUSDCx.address,
+                flowRate: inflowRateUsdc,
+                userData: ethers.utils.defaultAbiCoder.encode(["string"], ["bob"]),
+            }).exec(aliceSigner);
+
+            // Alice gets 98% because of refferal fee
+            expect(
+                (await launchpad.getIDAShares(ETHX_SUBSCRIPTION_INDEX, aliceSigner.address)).toString()
+            ).to.equal(`true,false,980000,0`);
+            // Admin and Carl split 2% of the shares bc of the 50% referral fee
+            expect(
+                (await launchpad.getIDAShares(ETHX_SUBSCRIPTION_INDEX, adminSigner.address)).toString()
+            ).to.equal(`true,false,10000,0`);
+            expect(
+                (await launchpad.getIDAShares(ETHX_SUBSCRIPTION_INDEX, bobSigner.address)).toString()
+            ).to.equal(`true,false,10000,0`);
```

Comment:
Can you add to this, close the stream and add expects for: 
- confirming the affiliate shares go to 0 on stream close
- open a stream without a referral and confirm it doesn't have an affiliate (DAO gets all the feeshares)

Reply:
@mikeghen no problem, changes made!